lib/ClangImporter/ImportType.cpp

```diff
@@ -716,6 +716,9 @@ namespace {
       // the hint, and if the typedef uses different bridgeability than the
       // context then we may also need to bypass the typedef.
       } else {
+        if (type->getDecl()->getName() == "BOOL")
+          return { mappedType, ImportHint::BOOL };
```

Comment:
Shouldn't this be covered by the "special" typedefs above?

Also, should BOOLEAN also get this treatment?

Reply:
That is what I was looking at initially and wanted to figure out how to deal with.  That iterates a `SpecialTypedefNames` table which is populated using `CFPointeeInfo::classifyTypedef`.  I suppose that I can add a special case there for the `BOOL` type, just wasn't sure if that was preferable to this.

Reply:
That's not right. CFPointeeInfo is for CF types. The thing you actually want is MappedTypes.def, although I don't know how that'll handle having two entries named "BOOL".

Reply:
Hmm, I'm not aware of any `BOOLEAN` type in Windows, and I don't seem to see that on MSDN, so, lets skip that for now.

However, the problem now is ObjC - the interaction between Windows and ObjC is an interesting corner case - ObjC will resolve `BOOL` by finding the typedef for `BOOL`.  This means that if `objc.h` is included first, `BOOL` will be `signed char` or `_Bool` based on the bitness.  If `Windows.h` is imported first, it will be `int`.  Because we do a single pass, we either see the mapping for `BOOL` to `ObjectiveC.ObjCBool` or `WinSDK.WindowsBOOL`.  I think that we may need to special case the `BOOL` handling.

Reply:
I think that I have an idea of how to handle this.  Its kinda ugly in that it requires manual hand holding to bridge, but that is fine.  In general it amounts to:

- **ALWAYS** Objective-C Bool (`ObjCBool`)
- If Windows: flip to Windows Bool (`WindowsBOOL`)
- If ObjC interop is enabled and the canonical type match the ObjC boolean type flip to Objective-C Bool (`ObjCBool`)

This should allow proper behaviour on all the targets and the various combinations I believe.

Reply:
That sounds about right. If you wanted to make it simpler, you could say `-enable-objc-interop` is always the deciding factor.

Reply:
I found `BOOLEAN` in the list at https://docs.microsoft.com/en-us/windows/desktop/winprog/windows-data-types, which is what I got when I searched "Windows boolean type", but maybe it's not used in practice?

Reply:
I think that the complex case is nicer, it bridges much more precisely.  However, you are correct!  There is a `BOOLEAN` type, and that is in use in Foundation even!  I think I will bridge that as well.  I want to get this in and then do that in a follow up though.

Reply:
Relevant reading:  
https://devblogs.microsoft.com/oldnewthing/20041222-00/?p=36923